flink-runtime/src/main/scala/org/apache/flink/runtime/taskmanager/TaskManager.scala

```diff
@@ -411,6 +411,23 @@ class TaskManager(
             log.debug(s"Cannot find task to fail for execution ${executionID})")
           }
 
+        // stops a task
+        case StopTask(executionID) =>
+          val task = runningTasks.get(executionID)
+          if (task != null) {
+            try {
+              task.stopExecution()
```

Comment:
Hmm, if I followed the control flow correctly, then the `stop` operation of the `SourceStreamTask` calls the `cancel` method of the `StreamSource`. The `StreamSource` calls `cancel` on the userFunction of type `SourceFunction`. Thus, all `SourceFunction` implementations have to be non-blocking when calling `cancel`. I haven't checked all of them but I've found the `KafkaSource` which also calls `shutdown` on a consumer which is of type `ZooKeeperConsumerConnector`. What I want to show you with my little trip through the class dependencies is that it's easy to get things wrong at one or the other function call. Moreover, the `Stoppable` interface never says that `stop` should be a non-blocking operation. Thus, I'm wondering how you know and can guarantee that it's non-blocking.